### PR TITLE
[JUJU-1087] Keep state worker alive for model life watcher failures

### PR DESCRIPTION
**Pull request by manadart**
Under #13633 we gave some leeway in the state worker for individual model workers that fail their pings to state.

Unfortunately this change was immaterial to the worker resiliency because the top level model life worker, which is added to the worker's catacomb, will kill the worker if it's session closes anyway.

Here we decouple it so that it is restarted whenever it completes. This is safe, because we can rely on the model sub-workers to kill the worker when failure to access Mongo causes repeated ping failures. Even if there are no workload models, there is still a controller model.

There is some test suite rearrangement in the second-to-last commit. The main change here is represented in the last one.

This is hard to unit test using the package's test infrastructure because destroying or resetting the test Mongo, causes a new one to come up on a different port and we can't cause the watcher to restart (it attempts to dial the old port). However, the tests do verify that ping failures bounce the worker in this case.

## QA steps

- Bootstrap and `juju enable-ha`, waiting for quiescence.
- Ensure that controller model logging config includes `juju.worker.state=DEBUG`.
- Connect to the Mongo primary and run `rs.stepDown()`.
- Run `juju debug-log -m controller` and look for something like:
```
machine-0: 16:29:01 INFO juju.worker.state restarting model watcher for error: state has been closed
```
- Now add and destroy a model. The logs should show that the life watcher is still working, indicated by entries such as:
```
machine-0: 16:29:43 DEBUG juju.worker.state model "54e659c1-ad2d-4f86-88fe-e63b1e7152b1" is dead
```

## Documentation changes

None.

## Bug reference

N/A

### CODE REVIEW

**Review by hmlanigan — APPROVED** [review]
Works as described.

File: worker/state/manifold.go [hunk 1]
```diff
@@ -52,9 +52,10 @@ func (config ManifoldConfig) Validate() error {
 
 const defaultPingInterval = 15 * time.Second
 
-// Manifold returns a manifold whose worker which wraps a
-// *state.State, which is in turn wrapper by a StateTracker.  It will
-// exit if the State's associated mongodb session dies.
+// Manifold returns a manifold whose worker which wraps a *state.State,
+// which is in turn wrapper by a StateTracker.
```
Comment (wallyworld) [inline worker/state/manifold.go:56]: wrapped

**Comment by manadart** [comment]
$$merge$$

**Comment by manadart** [comment]
!!build!!

**Comment by manadart** [comment]
!!build!!

**Comment by manadart** [comment]
!!build!!

**Comment by manadart** [comment]
Closing this one, as there is a lot of associated risk with the change. It can serve as a thought experiment, and potentially revisited in the future.

**Pull request closed without merging**